core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -173,6 +174,24 @@ private[scheduler] class BlacklistTracker (
         listenerBus.post(SparkListenerExecutorBlacklisted(now, exec, newTotal))
         executorIdToFailureList.remove(exec)
         updateNextExpiryTime()
+        // Add executor to blacklist before attempting to kill it. This allows a scheduler backend
+        // to immediately fail to allocate resources on this executor, since killing could be
+        // asynchronous.
```

Comment:
We chatted offline and there were two misunderstandings (on my end) here. One is that the order of killing and adding to the blacklist in `BlacklistTracker` doesn't matter since that code is holding the taskScheduler lock. The other is that `scheduler.nodeBlacklist.contains(executorId))` in CoarseGrainedSchedulerBackend was wrong - it should be `scheduler.nodeBlacklist.contains(hostname))`.

The next patch set that I upload will address both these issues.